File: LegacySource/settlement/room/law/guard/GuardPower.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Settlement.Room.Law.Guard
{
    public class GuardPower : SAVABLE
    {
        private Div d;
        private readonly DIV_SPEC spec = new DIV_SPEC();

        public GuardPower()
        {
            spec.Race = () => d.Race();
            spec.Training = (StatTraining tr) => tr.Stat.Div().GetD(d);
            spec.Equip = (EquipBattle e) => Math.Clamp((double)e.Stat().Div().Get(d) / STATS.POP().Pop(HTYPES.SOLDIER(), d), 0, 1);
            spec.Men = () => STATS.POP().Pop(HTYPES.GUARD(), d);
            spec.Faction = () => FACTIONS.Player();
            spec.Experience = () => STATS.BATTLE().COMBAT_EXPERIENCE.Div().GetD(d);
            spec.Name = () => null;
            spec.BannerI = () => 0;
        }

        private double res = 0;
        private double resD = 0;
        private int upI = -1;
        private int DI = 0;

        public double Get()
        {
            if (upI == GAME.UpdateI())
                return res;
            upI = GAME.UpdateI();
            if (DI >= GAME.ARMIES().Player().Divisions().Count)
            {
                DI = 0;
                res = resD;
                resD = 0;
            }

            Div d = GAME.ARMIES().Player().Divisions()[DI];
            if (SETT.ROOMS().GUARD.ActiveDuty.Is(d) && STATS.POP().Pop(HTYPES.GUARD(), d) > 0)
            {
                this.d = d;
                resD += GAME.Battle().Power.Get(spec);
            }
            DI++;

            return res;
        }

        public void Save(FilePutter file)
        {
            file.d(res);
        }

        public void Load(FileGetter file)
        {
            res = file.d();
        }

        public void Clear()
        {
            res = 0;
            resD = 0;
        }
    }
}

File: settlement/room/law/guard/GuardPower.java
package settlement.room.law.guard;

import java.io.IOException;

import game.GAME;
import game.battle.div.Div;
import game.battle.util.DIV_SPEC;
import game.faction.FACTIONS;
import game.faction.Faction;
import init.race.Race;
import init.type.HTYPES;
import settlement.main.SETT;
import settlement.stats.STATS;
import settlement.stats.colls.StatsBattle.StatTraining;
import settlement.stats.equip.EquipBattle;
import snake2d.util.file.FileGetter;
import snake2d.util.file.FilePutter;
import snake2d.util.file.SAVABLE;
import snake2d.util.misc.CLAMP;

public class GuardPower implements SAVABLE{

	
	private Div d;
	private final DIV_SPEC spec = new DIV_SPEC() {
		
		@Override
		public Race race() {
			return d.race();
		}
		
		@Override
		public double training(StatTraining tr) {
			return tr.stat.div().getD(d);
		}
		
		@Override
		public double equip(EquipBattle e) {
			return CLAMP.d((double)e.stat().div().get(d)/STATS.POP().pop(HTYPES.SOLDIER(), d), 0, 1);
		}
		
		@Override
		public int men() {
			return STATS.POP().pop(HTYPES.GUARD(), d);
		}
		
		@Override
		public Faction faction() {
			return FACTIONS.player();
		}
		
		@Override
		public double experience() {
			return STATS.BATTLE().COMBAT_EXPERIENCE.div().getD(d);
		}

		@Override
		public CharSequence name() {
			// TODO Auto-generated method stub
			return null;
		}

		@Override
		public int bannerI() {
			// TODO Auto-generated method stub
			return 0;
		}
	};
	
	private double res = 0;
	private double resD = 0;
	private int upI = -1;
	private int DI = 0;
	
	public double get() {
		if (upI == GAME.updateI())
			return res;
		upI = GAME.updateI();
		if (DI >= GAME.ARMIES().player().divisions().size()) {
			DI = 0;
			res = resD;
			resD = 0;
		}
		
		Div d = GAME.ARMIES().player().divisions().get(DI);
		if (SETT.ROOMS().GUARD.activeDuty.is(d) && STATS.POP().pop(HTYPES.GUARD(), d) > 0) {
			this.d = d;
			resD += GAME.battle().power.get(spec);
		}
		DI++;
		
		return res;
	}

	@Override
	public void save(FilePutter file) {
		file.d(res);
	}

	@Override
	public void load(FileGetter file) throws IOException {
		res = file.d();
	}

	@Override
	public void clear() {
		res = 0;
		resD = 0;
	}
	
}
